File: mvvm/src/main/java/com/lookballs/mvvm/BaseFragmentPagerAdapter.java
package com.lookballs.mvvm;

import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import androidx.fragment.app.FragmentActivity;
import androidx.fragment.app.FragmentManager;
import androidx.fragment.app.FragmentPagerAdapter;
import androidx.viewpager.widget.ViewPager;

import java.util.ArrayList;
import java.util.List;

/**
 * FragmentPagerAdapter基类
 */
public final class BaseFragmentPagerAdapter<F extends Fragment> extends FragmentPagerAdapter {

    /**
     * Fragment 集合
     */
    private final List<F> mFragments = new ArrayList<>();
    /**
     * Fragment 标题
     */
    private final List<CharSequence> mFragmentTitles = new ArrayList<>();
    /**
     * 当前显示的Fragment
     */
    private F mShowFragment;
    /**
     * 当前 ViewPager
     */
    private ViewPager mViewPager;
    /**
     * 设置成懒加载模式
     */
    private boolean mLazyMode = true;

    public BaseFragmentPagerAdapter(@NonNull FragmentActivity activity) {
        this(activity.getSupportFragmentManager());
    }

    public BaseFragmentPagerAdapter(@NonNull Fragment fragment) {
        this(fragment.getChildFragmentManager());
    }

    public BaseFragmentPagerAdapter(@NonNull FragmentManager manager) {
        super(manager, BEHAVIOR_RESUME_ONLY_CURRENT_FRAGMENT);
    }

    @NonNull
    @Override
    public F getItem(int position) {
        return mFragments.get(position);
    }

    @Override
    public long getItemId(int position) {
        return getItem(position).hashCode();
    }

    @Override
    public int getCount() {
        return mFragments.size();
    }

    @Nullable
    @Override
    public CharSequence getPageTitle(int position) {
        return mFragmentTitles.get(position);
    }

    @SuppressWarnings("unchecked")
    @Override
    public void setPrimaryItem(@NonNull ViewGroup container, int position, @NonNull Object object) {
        super.setPrimaryItem(container, position, object);
        if (getShowFragment() != object) {
            // 记录当前的Fragment对象
            mShowFragment = (F) object;
        }
    }

    /**
     * 添加 Fragment
     */
    public void addFragment(@NonNull F fragment) {
        addFragment(fragment, null);
    }

    public void addFragment(@NonNull F fragment, CharSequence title) {
        mFragments.add(fragment);
        mFragmentTitles.add(title);
        refreshAdapter();
    }

    public void addFragments(List<F> fragments) {
        addFragments(fragments, null);
    }

    public void addFragments(List<F> fragments, List<CharSequence> titles) {
        mFragments.clear();
        mFragmentTitles.clear();
        mFragments.addAll(fragments);
        mFragmentTitles.addAll(titles);
        refreshAdapter();
    }

    /**
     * 移除Fragment
     */
    public void removeFragment(int index) {
        if (mFragments.size() == 0 || index >= mFragments.size()) {
            return;
        }
        mFragments.remove(index);
        mFragmentTitles.remove(index);
        refreshAdapter();
    }

    public void removeFragment(@NonNull F fragment) {
        int index = -1;
        for (int i = 0; i < mFragments.size(); i++) {
            if (mFragments.get(i) == fragment) {
                index = i;
                break;
            }
        }
        if (index != -1) {
            removeFragment(index);
        }
    }

    /**
     * 刷新适配器
     */
    private void refreshAdapter() {
        if (mViewPager != null) {
            notifyDataSetChanged();
            if (mLazyMode) {
                mViewPager.setOffscreenPageLimit(getCount());
            } else {
                mViewPager.setOffscreenPageLimit(1);
            }
        }
    }

    /**
     * 获取当前的Fragment
     */
    public F getShowFragment() {
        return mShowFragment;
    }

    /**
     * 获取某个 Fragment 的索引（没有就返回 -1）
     */
    public int getFragmentIndex(Class<? extends Fragment> clazz) {
        if (clazz != null) {
            for (int i = 0; i < mFragments.size(); i++) {
                if (clazz.getName().equals(mFragments.get(i).getClass().getName())) {
                    return i;
                }
            }
        }
        return -1;
    }

    @Override
    public void startUpdate(@NonNull ViewGroup container) {
        super.startUpdate(container);
        if (container instanceof ViewPager) {
            // 记录绑定 ViewPager
            mViewPager = (ViewPager) container;
            refreshLazyMode();
        }
    }

    /**
     * 设置懒加载模式
     */
    public void setLazyMode(boolean lazy) {
        mLazyMode = lazy;
        refreshLazyMode();
    }

    /**
     * 刷新加载模式
     */
    private void refreshLazyMode() {
        if (mViewPager == null) {
            return;
        }

        if (mLazyMode) {
            // 设置成懒加载模式（也就是不限制 Fragment 展示的数量）
            mViewPager.setOffscreenPageLimit(getCount());
        } else {
            mViewPager.setOffscreenPageLimit(1);
        }
    }
}

File: mvvm-kt/src/main/java/com/lookballs/mvvm/BaseFragmentPagerAdapter.kt
package com.lookballs.mvvm

import android.view.ViewGroup
import androidx.fragment.app.Fragment
import androidx.fragment.app.FragmentActivity
import androidx.fragment.app.FragmentManager
import androidx.fragment.app.FragmentPagerAdapter
import androidx.viewpager.widget.ViewPager

/**
 * FragmentPagerAdapter基类
 */
class BaseFragmentPagerAdapter<F : Fragment>(manager: FragmentManager) : FragmentPagerAdapter(manager, BEHAVIOR_RESUME_ONLY_CURRENT_FRAGMENT) {
    /**
     * Fragment 集合
     */
    private val mFragments: MutableList<F> by lazy {
        mutableListOf<F>()
    }

    /**
     * Fragment 标题
     */
    private val mFragmentTitles: MutableList<CharSequence?> by lazy {
        mutableListOf<CharSequence?>()
    }

    /**
     * 获取当前的Fragment
     */
    /**
     * 当前显示的Fragment
     */
    var showFragment: F? = null
        private set

    /**
     * 当前 ViewPager
     */
    private var mViewPager: ViewPager? = null

    /**
     * 设置成懒加载模式
     */
    private var mLazyMode = true

    constructor(activity: FragmentActivity) : this(activity.supportFragmentManager) {}
    constructor(fragment: Fragment) : this(fragment.childFragmentManager) {}

    override fun getItem(position: Int): Fragment {
        return mFragments[position]
    }

    override fun getItemId(position: Int): Long {
        return getItem(position).hashCode().toLong()
    }

    override fun getCount(): Int {
        return mFragments.size
    }

    override fun getPageTitle(position: Int): CharSequence? {
        return mFragmentTitles[position]
    }

    override fun setPrimaryItem(container: ViewGroup, position: Int, `object`: Any) {
        super.setPrimaryItem(container, position, `object`)
        if (showFragment !== `object`) {
            // 记录当前的Fragment对象
            showFragment = `object` as F
        }
    }

    /**
     * 添加 Fragment
     */
    @JvmOverloads
    fun addFragment(fragment: F, title: CharSequence? = null) {
        mFragments.add(fragment!!)
        mFragmentTitles.add(title)
        refreshAdapter()
    }

    @JvmOverloads
    fun addFragments(fragments: List<F>, titles: List<CharSequence>? = null) {
        mFragments.clear()
        mFragmentTitles.clear()
        mFragments.addAll(fragments!!)
        titles?.let {
            mFragmentTitles.addAll(it)
        }
        refreshAdapter()
    }

    /**
     * 移除Fragment
     */
    fun removeFragment(index: Int) {
        if (mFragments.size == 0 || index >= mFragments.size) {
            return
        }
        mFragments.removeAt(index)
        mFragmentTitles.removeAt(index)
        refreshAdapter()
    }

    fun removeFragment(fragment: F) {
        var index = -1
        for (i in mFragments.indices) {
            if (mFragments[i] === fragment) {
                index = i
                break
            }
        }
        if (index != -1) {
            removeFragment(index)
        }
    }

    /**
     * 刷新适配器
     */
    private fun refreshAdapter() {
        mViewPager?.apply {
            notifyDataSetChanged()
            offscreenPageLimit = if (mLazyMode) {
                count
            } else {
                1
            }
        }
    }

    /**
     * 获取某个 Fragment 的索引（没有就返回 -1）
     */
    fun getFragmentIndex(clazz: Class<out Fragment>): Int {
        for (i in mFragments.indices) {
            if (clazz.name == mFragments[i].javaClass.name) {
                return i
            }
        }
        return -1
    }

    override fun startUpdate(container: ViewGroup) {
        super.startUpdate(container)
        if (container is ViewPager) {
            // 记录绑定 ViewPager
            mViewPager = container
            refreshLazyMode()
        }
    }

    /**
     * 设置懒加载模式
     */
    fun setLazyMode(lazy: Boolean) {
        mLazyMode = lazy
        refreshLazyMode()
    }

    /**
     * 刷新加载模式
     */
    private fun refreshLazyMode() {
        if (mLazyMode) {
            // 设置成懒加载模式（也就是不限制 Fragment 展示的数量）
            mViewPager?.offscreenPageLimit = count
        } else {
            mViewPager?.offscreenPageLimit = 1
        }
    }
}